### ISSUE TITLE
Permission denied for chromedriver

### ISSUE THREAD

**Issue by sunny9495-dev** [ultrafunkamsterdam/undetected-chromedriver#537]
Unable to use chromedriver in multithread environment.

Getting the following error

`PermissionError: [Errno 13] Permission denied: 'C:\\Users\\Gaggu\\appdata\\roaming\\undetected_chromedriver\\chromedriver.exe'`

```
if __name__ == '__main__':
   # download_driver()
   patcher = webdriver.Patcher()
   lock_file(patcher.executable_path, lock=False)
   patcher.auto()
   lock_file(patcher.executable_path)
   webdriver.Patcher.is_binary_patched = lambda self: True
```

I use the following code and it seems `patcher.auto` is giving issue.
This is the complete error code

```
File "undetected_chromedriver\patcher.py", line 121, in auto
File "undetected_chromedriver\patcher.py", line 175, in unzip_package
File "zipfile.py", line 1616, in extract
File "zipfile.py", line 1687, in _extract_member
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\Gaggu\\appdata\\roaming\\undetected_chromedriver\\chromedriver.exe'
```
Any help would be appreciated

**Comment by sebdelsol** [comment]
You don't have permission on the chromedriver.exe file, the `lock_file()` I gave in another comment should make it writable. Could you check you chromedriver.exe and its folder permission ?

**Comment by sebdelsol** [comment]
The [link](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/528#issuecomment-1059812278) you provided above is meant to force the `executable_path` (i.e the chromedriver.exe path) in a unique tmp directory so that you won't have concurrency issues when running several scripts (or threads) that might try to patch the driver at the same time. 

The [code](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/528#issuecomment-1059812278) you referred in your first comment is meant to do that with another trick (patch the driver once at the beginning and prevent any further instantiation of `Chrome` from patching chromedriver.exe) but it works _only_ if you instantiate your Chrome objects concurrently in the same script (with threads or processes).

_Both are incompatible_ you have to chose one or the other depending on your needs or your preference.

**Comment by sunny9495-dev** [comment]
ok @sebdelsol, i prefer [this one](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/528#issuecomment-1059812278)

But the only issue, it is not downloading chromedriver in new machines, any way to download new chromedriver before we start threads

**Comment by sebdelsol** [comment]
It seems you don't have permission on the `"~/appdata/roaming/undetected_chromedriver"` directory. Maybe you're running your script as a user without administrator privilege.

**Comment by sunny9495-dev** [comment]
yes, i have admin access, iam trying the same in RDP too, but both are failing to download the chromedriver when using this [code](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/528#issuecomment-1059812278)

**Comment by sebdelsol** [comment]
check your "~/appdata/roaming/undetected_chromedriver" directory permissions.

**Comment by sunny9495-dev** [comment]
@sebdelsol, i have admin access mate, but here the issue is download the chromedriver, using this [code](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/528#issuecomment-1059812278), we create a tmp directory with chromedriver in it, but when i compile and run in new machines, where there is no undetected_chromedriver folder in appdata folder,it isnt runnning, as there undetected_chromedriver in appdata folder,  all i just want is download the undetected_chromedriver, when i tried this [code](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/479#issuecomment-1030814150) it is download the chromedriver successfully without any issue mate.

**Comment by sunny9495-dev** [comment]
found solution @sebdelsol i use `patcher = webdriver.Patcher()` `patcher.auto()` from this [code](https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/479#issuecomment-1030814150)

```
import os
import shutil
import tempfile
import time
import undetected_chromedriver as webdriver


def force_patcher_to_use(directory):
    # copy the chromedriver in directory
    exe = webdriver.Patcher.exe_name
    src = os.path.join(webdriver.Patcher.data_path, exe)
    executable_path = os.path.join(directory, exe)
    shutil.copyfile(src, executable_path)

    # monkey patch the Patcher class
    class PatcherWithForcedExecutablePath(webdriver.Patcher):
        def __init__(self, *args, **kwargs):
            kwargs["executable_path"] = executable_path
            super().__init__(*args, **kwargs)

    webdriver.Patcher = PatcherWithForcedExecutablePath

    return executable_path


if __name__ == "__main__":
    # create a tmp dir where the driver will be put
    patcher = webdriver.Patcher()
    patcher.auto()
    tmp_directory = os.path.normpath(tempfile.mkdtemp())
    executable_path = force_patcher_to_use(tmp_directory)

    driver = webdriver.Chrome(executable_path=executable_path)

    # some cleanup when done
    driver.quit()
    time.sleep(3)
    os.remove(executable_path)
    os.rmdir(tmp_directory)
```

**Comment by sebdelsol** [comment]
ah you're right it won't work if the `chromedriver.exe` is not already there because it tries to copy it in the temp file... 
Your solution won't work if you try to launch your script several time concurrently : you'll end up with a random permission error because of the `patcher.auto()` call.

**Comment by sunny9495-dev** [comment]
Oh.. any solution for it mate

**Comment by sebdelsol** [comment]
this works but it's very hacky :sweat_smile: 
```python
def force_patcher_to_use(directory):
    # copy the chromedriver in directory
    exe = webdriver.Patcher.exe_name
    executable_path = os.path.join(directory, exe)
    src = os.path.join(webdriver.Patcher.data_path, exe)
    if os.path.exists(src):
        shutil.copyfile(src, executable_path)

    # monkey patch the Patcher class
    class PatcherWithForcedExecutablePath(webdriver.Patcher):
        def __init__(self, *args, **kwargs):
            kwargs["executable_path"] = executable_path
            super().__init__(*args, **kwargs)

            # download the chromedriver if it was not already done above
            if not os.path.exists(executable_path):
                release = self.fetch_release_number()
                self.version_main = release.version[0]
                self.version_full = release
                self.unzip_package(self.fetch_package())

    webdriver.Patcher = PatcherWithForcedExecutablePath

    return executable_path
```
the better solution would be to fix undetected driver to accept an executable path...

**Comment by sunny9495-dev** [comment]
Thanks for the solution @sebdelsol, It works fine

**Comment by sebdelsol** [comment]
It's not ideal because it has to download the driver each time...

**Comment by sunny9495-dev** [comment]
Oh ok @sebdelsol, i hope the developer give an option to let choose the `executable_path` of `chromedriver` in next update

**Comment by sebdelsol** [comment]
the option already exists but is not forwarded to the patcher, it's one-liner [here](https://github.com/ultrafunkamsterdam/undetected-chromedriver/blob/a4cc4a8b72691a3f5462b6f23973727e2b03da9e/undetected_chromedriver/__init__.py#L208)

**Comment by ultrafunkamsterdam** [comment]
fixed in latest, where it randomizes the driver binary name, and also fixed passing the exact path
